---
title: "OS stratification by reclassified mutation status"
author: Thinh Tran
output: html_document
date: "2024-03-06"
---

# Introduction

This R script accompanies the manuscript *Machine learning cancer driver mutation predictions are valid in real-world data* (Tran et al., 2024). It performs survival analysis using weighted Cox proportional hazards models to assess the impact of gene mutations on overall survival (OS) in a cohort of cancer patients. The analysis leverages a set of clinical data, a mutation annotation format (MAF) file, and optionally, a gene alteration matrix (GAM). The primary goal is to investigate the relationship between mutational status and survival outcomes, using inverse probability of treatment weighting (IPTW) to adjust for potential confounders. Additionally, it compares the effect of oncogenic versus benign mutations on patient survival.

## Key functions
`weightedSurvivalGeneList`: A function that runs the weighted Cox PH models for each gene in a specified list.
`createPlotDFs`: A utility function to format the Cox PH results for easy plotting.

## How to run the code:
Set up and load data: Load the clinical data, MAF file, and optional GAM using fread() from the data.table package. You must also define the variables to be used for weighting.

Run Cox PH Models: Define the genes of interest and use the weightedSurvivalGeneList function to stratify outcomes based on mutational status. This will run a series of Cox PH models for each gene and compile the results.

Plot Results: Use ggplot2 to create a bubble plot that visualizes the hazard ratios (HRs) for oncogenic mutations versus no mutation, categorized by different VEPs.

Run Control Models: Compare patients with reclassified benign mutations versus no mutation, using the same workflow as for oncogenic mutations.

Calculate Relative Risk (RR): For a gene of interest, calculate the RR of survival between oncogenic and benign mutations using the results of the previous analyses.

# Codes

## Set-up

```{r setup, include=FALSE, eval=F}
library(data.table)
library(here)
library(survival)
library(survminer)
library(tidytable)
library(WeightIt)
library(cobalt)
library(cowplot)
library(survey)
library(stringdist)
library(stringr)
library(RColorBrewer)
library(tidytext)
library(corrplot)
library(ranger)
library(forcats)
library(ggplot2)

source("src/utils.R")

```

## Load data
This analysis requires two main files:

1. A clinical data file, containing patient-level clinical and demographic information, as well as outcome 

2. A MAF file containing mutation information and additional columns that annotate each variant using OncoKB as well as any additional VEP of interest.

A gene alteration matrix (GAM) is optional if you wish to identify a list of genes to run the OS analysis on based on % of alteration. This is a binary matrix with samples as rows and genes as columns. 

```{r load data, eval=F}
# Optional - binary gene alteration matrix
gam <- fread("data/msk_impact_nsclc_gam.RData")

# Clinical file
pt_subset <- fread("data/msk_impact_nsclc_clinical.csv")

# Annotated MAF
all_missense_process_extended <- fread("data/msk_impact_nsclc_missense_mutations.csv")

# a list of variables to be used in inverse probability of treatment weighting - these should be the exact column names present in the clinical data file
weighting_vars <-  c("AGE_AT_DX", "HAS_STAGE_IV_DX", "GENDER",
                     "RACE", "HISTOLOGY", "SMOKING",
                     "DIAGNOSIS_NUMBER_BINARY", "TMB_HIGH","MSI_HIGH" ,
                     "EXTERNAL_TREATMENT_LABEL",
                     grep("^METASTATIC_", colnames(pt_subset), value=T)))

```


## Run on IMPACT

We can now run a series of weighted Cox PH models to stratify outcomes (in this case overall survival) by mutational status. 

The underlying function that performs the IPTW and Cox PH regressions is `weightedSurvivalGeneList`, which can be modified in `utils.R`.

```{r IMPACT OS, eval=F}

# Define a list of columns to run on
# These columns will be the left-hand side of a series of Cox PH models
# Here, we include a column for binarized OncoKB status (oncogenic vs non-oncogenic) and reclassified status
# according to different VEPs
outcomes_rescue <- c("oncogenic_binary", "gene_of_interest",
                     grep("Rescue", colnames(all_missense_process_extended), value=T))

# Optional: if you have a binary GAM that can be used to calculate alteration frequency of the cohort to select frequently altered genes
# Here we selected genes altered in >= 2% of the cohort (including alterations by mutations, fusions and indels)
# if not, you can specify a list of gene names
# the genes run in this manuscript are below
# gene_to_run <- c("ALK", "APC","ARID1A","ARID2","ATM", "BRAF","CDKN2A",  "CTNNB1",  "EGFR",
#                  "ERBB2",   "KEAP1", "KMT2D" , "KRAS",    "MET",     "MGA",     "NF1",
#                  "NFE2L2",  "PIK3CA",  "PTEN",    "RB1" ,    "RBM10",   "RET",  "ROS1", "SETD2",  
#                  "SMAD4", "SMARCA4", "STK11",   "TERT",    "TP53", "U2AF1")

drivers <- gam[PATIENT_ID %in% pt_subset$PATIENT_ID, ]
drivers <- drivers[, colSums(drivers != 0) > nrow(pt_subset)*0.02, with = FALSE]
gene_to_run <- setdiff(colnames(drivers), c("PATIENT_ID", "Tumor_Sample_Barcode"))

# Run a list of CoxPH 
missense_surv_vus <- weightedSurvivalGeneList(maf=all_missense_process_extended, #genomic file
                                         clinical = pt_subset, #clinical file
                                         gene_to_run=gene_to_run, #list of genes to run
                                         outcomes=outcomes_rescue, #list of left-hand side columns to stratify by
                                         matching_vars=weighting_vars, #list of weighting variables
                                         survobjform = "Surv(DTE_TO_SEQUENCING, OS_DURATION, event = OS_STATUS)", #survival formula passed on to coxph(), left truncated here at sequencing date
                                         count_threshold = 10, #minimum number of patients in each strata to be included
                                         baseline = "other") # not control analysis

# Format result data.table for easy plotting
missense_surv_dfs <- createPlotDFs(missense_surv_vus,
                              gene_to_run = names(missense_surv_vus))

# When created bubble plot, we only plot the coefficients of reclassified pathogenic vs. no mutation, so these coefficients are subset here
impact_coefs_plot <- missense_surv_dfs$rescued[grepl("gene_of_interest|Pathogenic|Oncogenic", term),]

# Isolate genes with at least one significant coefficient
gene_with_coefs <- unique(impact_coefs_plot[!is.na(estimate_sig),]$gene)

```

## Make bubble plot
This code plots the hazard ratio of the reclassified pathogenic vs no mutation per gene per VEPs for IMPACT.

```{r plot, eval=F}

ggplot(impact_coefs_plot[gene %in% gene_with_coefs,], aes(y=ann_type, x=gene, fill=estimate_sig)) +
  geom_point(aes(size=strata_size, color=estimate_sig))+
  # geom_tile() +
  scale_fill_gradientn(colors = rev(hcl.colors(10, "RdBu")),
                       na.value = "#f5f5f5",
                       rescaler = ~ scales::rescale_mid(.x, mid = 1))+
  scale_color_gradientn(colors = rev(hcl.colors(10, "RdBu")),
                       na.value = "#f5f5f5",
                       rescaler = ~ scales::rescale_mid(.x, mid = 1))+
  scale_size(range=c(2, 12), breaks = c(50, 200, 1000, 2500))+ # for patient count
  labs(x="Gene", y="", color="OS HR", fill="OS HR", size = "Number of patients")+
  theme_minimal()+
  rotate_x_text(angle=90)+ 
  facet_grid(ann_type~., scales = "free", space = "free")+
  theme(
  strip.background = element_blank(),
  strip.text.y = element_blank())

```

## Run control model: reclassified benign vs no mutation

Now we will run a series of weighted Cox PH models to stratify outcomes (in this case overall survival) by mutational status, comparing patients with reclassified benign mutations vs no mutation in each gene.

This will use the same function `weightedSurvivalGeneList`, which has a `baseline` argument that subsets the data and sets the right baseline for the regressions.

```{r os control}
missense_surv_benignbaseline <- weightedSurvivalGeneList(maf=all_missense_process_extended,
                                         clinical = pt_subset,
                                         gene_to_run=gene_to_run,
                                         outcomes=outcomes_rescue,
                                         matching_vars=matching_vars_lung,
                                         survobjform = "Surv(DTE_TO_SEQUENCING, OS_DURATION, event = OS_STATUS)",
                                         count_threshold = 10,
                                         baseline = "benign")

missense_surv_benign_dfs <- createPlotDFs(missense_surv_benignbaseline,
                              gene_to_run = names(missense_surv_benignbaseline))

```

## Calculate relative risk metric

To quantified VEP performance in distinguishing VUSs with outcome implications by calculating the relative risk of survival (RR), defined for each gene as the ratio of OS HR for patients with reclassified oncogenic mutations vs. no mutation, compared to OS HR for patients with reclassified benign mutations vs. no mutation (Tunes da Silva et al. 2009). The higher the RR, the better a VEP at identifying mutations that affect outcomes. 

We will use the OS results from the previous two analyses to calculate the relative risk. First subset the result df with a gene of interest, then run the cell to calculate the RR and plot

```{r rr calculation}

gene_of_interest <- "KEAP1" # set your gene of interest here

gene_hr <- impact_coefs_plot[gene == gene_of_interest, ]
gene_hr_control <- missense_surv_benign_dfs$rescued[gene == gene_of_interest, ]

gene_hr_plot <- merge(gene_hr,
                      gene_hr_control,
                      by="ann_type",
                      suffixes = c("_reclassified_onc", "_reclassified_benign"))

# Calculate the RR
gene_hr_plot <- gene_hr_plot[,  hr_ratio := hr_reclassified_onc/hr_reclassified_benign] 

# Calculate the log standard error for the RR estimate
gene_hr_plot <- gene_hr_plot[, se_lnrt := sqrt((std.error_reclassified_onc)^2 + (std.error_reclassified_benign)^2)]

# Calculate the upper and lower 95% confidence interval
gene_hr_plot <- gene_hr_plot[, `:=`(ci_rt_upper = exp(log(hr_ratio)+1.96*se_lnrt),
                                    ci_rt_lower = exp(log(hr_ratio)-1.96*se_lnrt_IMPACT))]
# generate bar plot
ggplot(gene_hr_plot, aes(x=reorder(ann_type, -hr_ratio), y=hr_ratio))+
  geom_bar(stat = "identity", position="dodge")+
  geom_hline(yintercept = 1, color="red")+
  geom_errorbar(aes(ymin=ci_rt_lower, ymax=ci_rt_upper), width=.2,
                position=position_dodge(.9)) +
  theme_classic()+
  scale_y_continuous(expand=c(0,0))+
  rotate_x_text(angle=45)+
  labs(y=paste0("Relative risk of survival:", gene_of_interest,"\nreclassified oncogenic vs. reclassified benign"), x="Method")

```


